pkg/webreg/graphviz.go

```diff
@@ -2,183 +2,224 @@ package webreg
-		lastType = newType
 	}
-	// this is used to identify a node that can be linked from is this sg is a src for an edge
-	if lastType == nodeType {
-		sg.lastNode = lastIndex
-	} else {
-		sg.lastNode = mainGraph.subgraphs[lastIndex].lastNode
+	i := len(b.graph.subgraphs)
+	b.graph.subgraphs = append(b.graph.subgraphs, sg)
+	b.edge.srcType = subgraphType
+	b.edge.srcGraph = i
+	if hasSrc && edge != len(b.graph.edges) {
+		b.graph.edges[edge].dstGraph = i
 	}
-	if !isRoot {
-		mainGraph.subgraphs = append(mainGraph.subgraphs, sg)
-	}
-	return mainGraph, len(mainGraph.subgraphs) - 1
+	return i
 }
 
-func addEmptySubgraph(mainGraph graph, name string) (graph, int) {
-	sg := subgraph{
-		label:     name,
-		firstNode: -1,
-	}
-	emptyNode := node{label: "Intentionally left blank"}
-	mainGraph.nodes = append(mainGraph.nodes, emptyNode)
-	sg.nodes = append(sg.nodes, len(mainGraph.nodes)-1)
-	sg.firstNode = len(mainGraph.nodes) - 1
-	sg.lastNode = sg.firstNode
-	mainGraph.subgraphs = append(mainGraph.subgraphs, sg)
-	return mainGraph, len(mainGraph.subgraphs) - 1
-
+// addNode creates a single leaf node and, if necessary, an edge
+func (b *graphBuilder) addNode(sg *subgraph, n node) {
+	i := len(b.graph.nodes)
+	b.graph.nodes = append(b.graph.nodes, n)
+	sg.nodes = append(sg.nodes, i)
+	b.addEdge(i)
 }
 
-func workflowDotFile(name string, workflows registry.WorkflowByName, chains registry.ChainByName, wfType string) string {
-	workflow := workflows[name]
-	mainGraph := graph{
-		label: fmt.Sprintf("%s \"%s\"", wfType, name),
+// addEdge creates an edge based on the current state to node `i`
+// `b.edge` is manipulated so it contains the necessary information for the
+// subsequent edge.
+func (b *graphBuilder) addEdge(i int) {
+	b.edge.dstNode = i
+	if b.edge.srcType != noEdge {
+		b.graph.edges = append(b.graph.edges, b.edge)
 	}
-	var preIndex, testIndex, postIndex int
-	if len(workflow.Pre) == 0 {
-		mainGraph, preIndex = addEmptySubgraph(mainGraph, "Pre")
-	} else {
-		mainGraph, preIndex = addSubgraph(mainGraph, "Pre", workflow.Pre, chains, false, false, false)
-	}
-	if len(workflow.Test) == 0 {
-		mainGraph, testIndex = addEmptySubgraph(mainGraph, "Test")
-	} else {
-		mainGraph, testIndex = addSubgraph(mainGraph, "Test", workflow.Test, chains, false, false, false)
-	}
-	if len(workflow.Post) == 0 {
-		mainGraph, postIndex = addEmptySubgraph(mainGraph, "Post")
-	} else {
-		mainGraph, postIndex = addSubgraph(mainGraph, "Post", workflow.Post, chains, false, false, false)
-	}
-	preTestEdge := edge{
-		src:     preIndex,
-		dst:     testIndex,
-		srcType: subgraphType,
-		dstType: subgraphType,
-	}
-	mainGraph.edges = append(mainGraph.edges, preTestEdge)
-	testPostEdge := edge{
-		src:     testIndex,
-		dst:     postIndex,
-		srcType: subgraphType,
-		dstType: subgraphType}
-	mainGraph.edges = append(mainGraph.edges, testPostEdge)
-	return writeDotFile(mainGraph)
+	b.edge = edge{srcType: nodeType, srcNode: b.edge.dstNode, dstType: nodeType}
```

Comment:
Assigning `srcNode: b.edge.dstNode` is a little bit misleading, better to change it into `srcNode: i` to underline that `i`, the node has just been added, is going to become the source for the next edge to come.

Reply:
It's supposed to reflect this comment (or vice versa), which is how I read it: https://github.com/openshift/ci-tools/pull/3189/files#diff-de4c46ee468a303e27ac25c79450556c326e7865b3dfaa6377220b41ee0852dfR80-R83.

```
Whenever a node (i.e. a step) is added to the graph, an edge is added
based on the current state of the variable.  Its destination is the
newly added node.  This variable is then changed in preparation for
the subsequent edge (its source is the one just added).
```